SanctumLabs/ujumbe
# Sms submitted consumer

## Description

This adds logic and implementation to handle consuming messages from the `sms_submitted_topic`, sending the SMS using Twilio 3rd Party service and publishing(or producing) an SMS_SENT event on successfully sending the message.

This adds the following:

1. `commit` and `commit_async` wrapper methods on the `KafkaConsumer` class to enable committing messages either synchronously or asynchronously to control sequential execution flow. This allows to only commit a message if there is a success, if there is a failure, the message is not committed and is retired by the consumer. This should be controlled reasonably & if the consumer fails on a number of retries the message should be handled appropriately.
2. `name` property on abstract classes to be used in log prefixes for easier log formatting
3. Update `.env.example` file with sample environment variables.


## How can we test this?

Ensure you have setup a `Twilio` account and followed the instructions to get an `AUTH_TOKEN` & `ACCOUNT_SID` and add those to an `.env` file in the root of the project. Instructions can be found [here](https://www.twilio.com/docs/sms/tutorials/how-to-send-sms-messages-python).

1. Run `make start-docker` to run all profiles in the `docker-compose.yml` file.
5. Run the main application with `make run-reload`
6. Run the `sms_received_consumer` with `python /app/workers/consumers/sms_received/__main__.py`
7. Run the `sms_submitted_consumer` with `python /app/workers/consumers/sms_submitted/__main__.py`
8. Make a HTTP request using the Swagger API on this [URL](http://127.0.0.1:5000/docs#/SMS/send_sms_api_v1_sms__post). You can optionally make a curl request or use any REST client:

``` shell
curl -X 'POST' \
  'http://127.0.0.1:5000/api/v1/sms/' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "sender": "<SENDER_NUMBER>",
  "recipient": "<RECIPIENT_NUMBER>",
  "message": "<YOUR MESSAGE>"
}'
```

The recipient should get the message you sent and the database should have stored both the response from Twilio and the SMS that is sent.

## Checklist:
